### PR TITLE
redboltz-mqtt_cpp: added flags for redboltz mqtt project + support conan v2

### PR DESCRIPTION
**Pull request by MateuszMiekicki**
Specify library name and version:  **redboltz-mqtt_cpp/13.1.0**

Hi, I noticed that this particular library is wrongly added comparing with the original repo, plus the flags for cmake are missing during the build. I've done this on the fly, taking examples from other projects. All comments welcome.

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by jwillikers — CHANGES_REQUESTED** [review]

**Comment by jwillikers** [comment]
> @jwillikers Thank you very much for your comments, I will try to improve everything in the final version. For the time being, I am struggling to create a package at all.
> 
> conan create . 13.1.0@
> 
> > redboltz-mqtt_cpp/13.1.0: Aggregating env generators
> > [HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] 'fPIC' option not found
> > [HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK
> > redboltz-mqtt_cpp/13.1.0: Calling build()
> > redboltz-mqtt_cpp/13.1.0: CMake command: cmake -G "Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:\Users\haven.conan\data\redboltz-mqtt_cpp\13.1.0__\source"
> > CMake Error: The source directory "C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/source" does not appear to contain CMakeLists.txt.
> > Specify --help for usage, or press the help button on the CMake GUI.
> > redboltz-mqtt_cpp/13.1.0:
> > redboltz-mqtt_cpp/13.1.0: ERROR: Package '5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9' build failed
> > redboltz-mqtt_cpp/13.1.0: WARN: Build folder C:\Users\haven.conan\data\redboltz-mqtt_cpp\13.1.0__\build\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
> > ERROR: redboltz-mqtt_cpp/13.1.0: Error in build() method, line 64
> > cmake.configure()
> > ConanException: Error 1 while executing cmake -G "Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:\Users\haven.conan\data\redboltz-mqtt_cpp\13.1.0__\source"
> 
> After running this command, unfortunately the package fails to build. Do you know where the problem lies? I know it can't find the CMakeList, but why? As far as I know, this library is in cmak as an interface, not as shared or static.

@MateuszMiekicki I think you're missing the necessary `source` method to actually grab the project's source code.

**Comment by MateuszMiekicki** [comment]
> > @jwillikers Thank you very much for your comments, I will try to improve everything in the final version. For the time being, I am struggling to create a package at all.
> > conan create . 13.1.0@
> > > redboltz-mqtt_cpp/13.1.0: Aggregating env generators
> > > [HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] 'fPIC' option not found
> > > [HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK
> > > redboltz-mqtt_cpp/13.1.0: Calling build()
> > > redboltz-mqtt_cpp/13.1.0: CMake command: cmake -G "Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:\Users\haven.conan\data\redboltz-mqtt_cpp\13.1.0__\source"
> > > CMake Error: The source directory "C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/source" does not appear to contain CMakeLists.txt.
> > > Specify --help for usage, or press the help button on the CMake GUI.
> > > redboltz-mqtt_cpp/13.1.0:
> > > redboltz-mqtt_cpp/13.1.0: ERROR: Package '5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9' build failed
> > > redboltz-mqtt_cpp/13.1.0: WARN: Build folder C:\Users\haven.conan\data\redboltz-mqtt_cpp\13.1.0__\build\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
> > > ERROR: redboltz-mqtt_cpp/13.1.0: Error in build() method, line 64
> > > cmake.configure()
> > > ConanException: Error 1 while executing cmake -G "Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/Users/haven/.conan/data/redboltz-mqtt_cpp/13.1.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:\Users\haven.conan\data\redboltz-mqtt_cpp\13.1.0__\source"
> > 
> > 
> > After running this command, unfortunately the package fails to build. Do you know where the problem lies? I know it can't find the CMakeList, but why? As far as I know, this library is in cmak as an interface, not as shared or static.
> 
> @MateuszMiekicki I think you're missing the necessary `source` method to actually grab the project's source code.

It helped, I was able to build what I wanted. I just don't know why it doesn't pass CI, I don't have access to the details.

**Comment by jwillikers** [comment]
@MateuszMiekicki It says too many blank lines are in `conandata.yml`.

**Comment by MateuszMiekicki** [comment]
> @MateuszMiekicki It says too many blank lines are in `conandata.yml`.

Thanks, it has been corrected

**Comment by MateuszMiekicki** [comment]
> @MateuszMiekicki It says too many blank lines are in `conandata.yml`.

Another problem xD Without your help I unfortunately cannot see the details.

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 1]
```diff
@@ -1,57 +1,97 @@
...
+    # exports_sources = "include/*"
     no_copy_source = True
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "CMakeDeps", "CMakeToolchain"
 
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
+    options = {
+        "cpp17": [True, False],
```
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:21]: Correct. @MateuszMiekicki please, remove it. Any configuration related to the supported c++ standard should be managed by settings.compiler.cppstd

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 2]
```diff
@@ -1,57 +1,97 @@
...
+        "fPIC": [True, False],
```
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:22]: @MateuszMiekicki Could you please check if this library offers shared/static support?

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 3]
```diff
@@ -1,57 +1,97 @@
...
+        "with_static_boost": [True, False],
+        "with_static_openssl": [True, False],
```
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:24]: Correct. It should be removed, please. All those options are part of dependencies options, not managed by this package.

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 4]
```diff
@@ -1,63 +1,115 @@
...
-                  destination=self._source_subfolder, strip_root=True)
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["MQTT_BUILD_EXAMPLES"] = False
+        tc.variables["MQTT_BUILD_TESTS"] = False
+        tc.variables["MQTT_ALWAYS_SEND_REASON_CODE"] = self.options.mqtt_always_send_reason_code
+        tc.variables["MQTT_USE_STATIC_BOOST"] = self.options.with_static_boost
+        if self.options.with_static_boost:
+            openssl = self.dependencies["boost"]
+            openssl.shared = True
```
Comment (uilianries) [inline recipes/redboltz-mqtt_cpp/all/conanfile.py:83]:
```suggestion
            openssl = self.dependencies["boost"]
            openssl.shared = True
```
It's read-only, you can't force it. Please, remove it.

**Review by uilianries — CHANGES_REQUESTED** [review]
Thank you for your contribution. Please, check the review, there are important points which are not allowed in CCI and should be removed/updated

File: recipes/redboltz-mqtt_cpp/all/conanfile.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MateuszMiekicki) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:21]: Isn't that enough?  https://github.com/conan-io/conan-center-index/blob/56f06c3fb4e56981ef088a5eaee25adbbd514ff6/recipes/redboltz-mqtt_cpp/all/conanfile.py#L69
Simply the project contains such a flag in CMakeLists and thanks to it we know in which version of c++ to build the project.

File: recipes/redboltz-mqtt_cpp/all/conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MateuszMiekicki) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:22]: https://github.com/redboltz/mqtt_cpp/blob/master/include/CMakeLists.txt#L7 The library is as header only

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 5]
```diff
@@ -1,57 +1,97 @@
...
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["MQTT_BUILD_EXAMPLES"] = False
+        tc.variables["MQTT_BUILD_TESTS"] = False
+        tc.variables["MQTT_ALWAYS_SEND_REASON_CODE"] = self.options.mqtt_always_send_reason_code
+        tc.variables["MQTT_USE_STATIC_BOOST"] = self.options.with_static_boost
```
[the remark this answers is not in the document]
Comment (MateuszMiekicki) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:80]: Actually looks better, corrected.

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 6]
```diff
@@ -1,57 +1,97 @@
...
+        tc.variables["MQTT_USE_STATIC_OPENSSL"] = self.options.with_static_openssl
+        tc.variables["MQTT_USE_TLS"] = self.options.with_tls
+        tc.variables["MQTT_USE_WS"] = self.options.with_websocket
+        if self.options.get_safe("cpp17"):
```
[the remark this answers is not in the document]
Comment (MateuszMiekicki) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:90]: This is where I would like to check the flag, the standard I check earlier

**Comment by MateuszMiekicki** [comment]
@jwillikers @uilianries Hey, can I have some comments so I can merge this?

File: recipes/redboltz-mqtt_cpp/all/conanfile.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/redboltz-mqtt_cpp/all/conanfile.py:23]: @MateuszMiekicki please use the template approach for cppstd version: https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py#L77

**Comment by MateuszMiekicki** [comment]
> Thank you for your contribution. Please, check the review, there are important points which are not allowed in CCI and should be removed/updated

Hi, could you point me to this place? Somehow it has eluded me.

**Comment by SSE4** [comment]
> Thank you for your contribution. Please, check the review, there are important points which are not allowed in CCI and should be removed/updated

@uilianries could you possible be more specific? what are these "important points"? what exactly should be "removed" or "updated"? it's a bit unclear what exactly do you mean...

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 7]
```diff
@@ -1,64 +1,139 @@
...
+    def _min_cppstd(self):
+        if self.options.enable_std_containers:
+            return "17"
+        return "14"
+
+    @property
+    def _minimum_compiler_version(self):
+        return {
+            "17": {
+                "Visual Studio": "16",
```
Comment (uilianries) [inline recipes/redboltz-mqtt_cpp/all/conanfile.py:52]:
```suggestion
                "Visual Studio": "16",
                "msvc": "1923",
```
Conan 2.x does not use Visual Studio, but msvc only as compiler name.

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 8]
```diff
@@ -1,64 +1,139 @@
...
+                "gcc": "7.3",
+                "clang": "6.0",
+                "apple-clang": "10.0"
+            },
+            "14": {
+                "Visual Studio": "15",
```
Comment (uilianries) [inline recipes/redboltz-mqtt_cpp/all/conanfile.py:58]:
```suggestion
                "Visual Studio": "15",
                "msvc": "1916",
```
Conan 2.x does not use Visual Studio, but msvc only as compiler name.

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 9]
```diff
@@ -1,64 +1,139 @@
...
+                "gcc": "6",
+                "clang": "3.4",
+                "apple-clang": "5.1"
+            },
+        }[self._min_cppstd]
+
 
     def requirements(self):
-        self.requires("boost/1.79.0")
+        self.requires("boost/1.81.0")
```
Comment (uilianries) [inline recipes/redboltz-mqtt_cpp/all/conanfile.py:67]:
```suggestion
        self.requires("boost/1.82.0")
```

File: recipes/redboltz-mqtt_cpp/all/conanfile.py [hunk 10]
```diff
@@ -1,64 +1,139 @@
...
-        self.cpp_info.set_property("cmake_file_name", "mqtt_cpp")
-        self.cpp_info.set_property("cmake_target_name", "mqtt_cpp::mqtt_cpp")
-
-        #  TODO: to remove in conan v2 once cmake_find_package_* generators removed
-        self.cpp_info.filenames["cmake_find_package"] = "mqtt_cpp"
-        self.cpp_info.filenames["cmake_find_package_multi"] = "mqtt_cpp"
-        self.cpp_info.names["cmake_find_package"] = "mqtt_cpp"
-        self.cpp_info.names["cmake_find_package_multi"] = "mqtt_cpp"
+        self.cpp_info.set_property("cmake_file_name", "mqtt_cpp_iface")
+        self.cpp_info.set_property("cmake_target_name", "mqtt_cpp_iface::mqtt_cpp_iface")
```
Comment (uilianries) [inline recipes/redboltz-mqtt_cpp/all/conanfile.py:116]:
```suggestion
        self.cpp_info.set_property("cmake_file_name", "mqtt_cpp_iface")
        self.cpp_info.set_property("cmake_target_name", "mqtt_cpp_iface::mqtt_cpp_iface")
```
Why did it change now? These targets were wrong or is just a new name from the latest version?

**Comment by uilianries** [comment]
@MateuszMiekicki Please, don't forget to sign the CLA attached to this PR. It's listed in the first comment of this PR.

**Pull request closed without merging**